### DISCUSSION TITLE
podman pod with phpmyadmin works ok using podman-compose but not using buildah

### DISCUSSION THREAD

**Discussion by barpasc** [containers/podman#20967] [General]
Hello,

I am trying to build a small home web development container/pod. 

From [this link](https://www.section.io/engineering-education/dockerized-php-apache-and-mysql-container-development-environment/) I can run apache2, php, mysql and phpmyadmin.  

I use `podman-compose up -d ` to run the server and access localhost:8000 and localhost:8080.

I found another script [from there](https://www.jakesco.com/posts/podman-lamp/) to do the same but I would like to add phpmyadmin to the created pod.

I tried the following

`
```
#!/bin/sh

echo -e "Building Apache PHP image"
id=$(buildah from --pull php:7.4.10-apache-buster)
buildah run $id docker-php-ext-install pdo pdo_mysql
buildah commit $id php-apache

echo -e "\nBuilding MySQL image"
id=$(buildah from --pull mysql:8)
buildah commit $id php-mysql

echo -e "\nBuilding phpmyadmin image"
id=$(buildah from --pull phpmyadmin:latest)
buildah commit $id php-pma

podman run \
    --detach \
    --pod new:php-lamp \
    --publish 8000:80 \
    --publish 8080:3306 \
    --name php-dev \
    --security-opt label=disable \
    --volume ./src:/var/www/html \
    php-apache

podman run \
    --detach \
    --pod php-lamp \
    --name mysql-dev \
    --env MYSQL_ROOT_PASSWORD=dev \
    --security-opt label=disable \
    --volume /home/barpasc/dev-web1/mysqldb:/var/lib/mysql:Z \
    --restart on-failure \
    php-mysql

podman run \
    --pod php-lamp \
    --name pma \
    --security-opt label=disable \
    -e PMA_HOST=127.0.0.1 \
    -e PMA_PORT=8080 \
   php-pma
```
`
and run it as user sh mklamp.sh

but I can only access index.php (http://127.0.0.1:8000/) but phpmyadmin (http://127.0.0.1:8080/) not accessible

`curl localhost:8080` 
returns _curl: (56) Recv failure: Connection reset by peer_

Any help on this?

**Comment by rhatdan** [comment]
--publish 8080:3306 \

Do you have something listening on 3306?

**Comment by barpasc** [comment]
--publish 8080:3306 \

I thought that was for mysql default port

--publish 8080:80

I tried the above and `http://127.0.0.1:8080/ `lands on the index.php file stored in src folder.

Any idea? I'm quite new to podman and I have basic level of tcp port

**Comment by barpasc** [comment]
ok a bit of research and I can run mysql with phpadmin these images using

```
podman pod create --name db0 --publish 8080:80 --publish 3306:3306

podman run --name ctnmysql --pod db0 -v /home/barpasc/dev-web1/mysqldb:/var/lib/mysql:Z -e MYSQL_ROOT_PASSWORD="MYSQL_ROOT_PASSWORD" -e MYSQL_USER="MYSQL_USER" -e MYSQL_PASSWORD="MYSQL_PASSWORD" -d php-mysql

podman run --name phpmyadmin0 --pod db0 -d -v /home/barpasc/dev-web1/phpmyadmin:/etc/phpmyadmin/config.user.inc.php:Z -e PMA_HOST=127.0.0.1 -e PMA_PORT=3306 php-pma
```

credits goes to the only answer on [this site](https://unix.stackexchange.com/questions/649190/how-can-i-run-mariadb-and-phpmyadmin-podman-containers-and-access-my-database-fr)

How can I now run the webserver? It seems something is not right with port numbers using 

```
podman pod create --name db0 --publish 8000:80 --publish 8080:80 --publish 3306:3306
podman run -d --name ctn_httpd_php --pod db0 --volume /home/barpasc/dev-web1/src:/var/www/html:Z httpd-php
```

ends up in degraded pod status and only phpinfo() page showing, not phpmyadmin login page as above

Is it possible to create containers for different apps port using podman? If not, is it possible to create multiple pods that can access the same ressources, like one pod with a  db container and phpmyadmin or pgadmin container and another pod with a webserver container and a container to the database in the first pod?

**Reply by rhatdan** [reply]
Would it have helped if we improved the man pages?  If yes, then would you consider opening a PR to fix the man pages.

**Reply by barpasc** [reply]
To be honest, I haven't taken time to look at the man pages...

I have to say I am new to containers, I was using virtualbox but my computer laptop freezes somewhere every 30-60 days, I keep a non gui vm that I use as a local web server all the time and sometimes windows 10 running... the temperature of either cpu, ram, nvme hd... might be the problem. System logs show some warning on temperature above threshold... but no additional information

I have never used docker and know very little about it but it requires an account and I would like to start learning on containers with something simple with minimum requirements. Moreover from what I know docker requires a system daemon and podman doesn't.

So I'm trying to see if lighter os process solves my laptop problem if not it might be a hardware issue (i customized some of the components...). So that's the whole story why I'm not getting too deep into system admin knowledge, it's way above my skills

**Reply by rhatdan** [reply]
Well welcome to the club, hope you enjoy your container experience.

**Comment by barpasc** [comment]
as this is my first time using containers, no docker experience at all, i understand podman doesn't allow multiports connection in the same container/pod, I'm not sure about what I'm saying, there is a lot of website with "Configuring container networking with Podman" but it's quite technical

Would writing a script that creates a pod with port 8000 for web php page and when database work is required a script that closes the port 8000 pod and create a pod that allow phpmyadmin, pgadmin... database work. A script that acts like a switch depending on task front end web dev tasks of back end web dev tasks? I am going to try that unless someone has the answer or experience on this

**Reply by rhatdan** [reply]
The Linux kernel does not allow two different processes to listen on the same port, by default.  Containers are just processes, so the same rule applies to processes within containers.

The rule of thumb should be that a single container controls the port, If you want to database and webserver to both listen on port 8000 then I think you have bad design.  They should be on different ports.  If you want to switch the port between different containers, I would say to have a relay port that listens to the incoming connections then forwards the message to the database or web server depending on the content of the message.  The  Web PHP page and Database would then listen on different ports potentially internally and not exposed on the external network

**Reply by barpasc** [reply]
My first question/port has mistakes in defining ports. I'll leave like that as there are answers below to it and it might help others to understand.

Starting from first using podman-compose up and then running podman pods with same containers as in podman-compose uses in yaml file made things confusing because with podman compose up it's possible to access different ports within the same process.
